src/bvals/comms/bnd_info.cpp

```diff
@@ -190,69 +199,129 @@ BndInfo BndInfo::GetSendBndInfo(std::shared_ptr<MeshBlock> pmb, const NeighborBl
-      out.var = v->data.Get();
-    } else if (nb.snb.level < mylevel) {
-      // "Same" logic is the same for loading to a coarse buffer, just using
-      // c_cellbounds
-      out.idxer[idx] = CalcLoadIndices<InterfaceType::FineToCoarse>(
-          nb.ni, el, {Nt, Nu, Nv}, pmb->c_cellbounds);
-      out.prores_idxer[static_cast<int>(el)] = out.idxer[idx];
-      out.refinement_op = RefinementOp_t::Restriction;
-      out.var = v->coarse_s.Get();
-    } else {
-      out.idxer[idx] = CalcLoadIndices<InterfaceType::CoarseToFine>(
-          nb.ni, el, {Nt, Nu, Nv}, pmb->cellbounds);
-      out.var = v->data.Get();
-    }
+    out.idxer[static_cast<int>(el)] =
+        CalcIndices(nb, pmb, el, IndexRangeType::Interior, true, {Nt, Nu, Nv});
   }
+  out.refinement_op = RefinementOp_t::Restriction;
   return out;
 }
 
-BndInfo BndInfo::GetSetBndInfo(std::shared_ptr<MeshBlock> pmb, const NeighborBlock &nb,
-                               std::shared_ptr<Variable<Real>> v,
-                               CommBuffer<buf_pool_t<Real>::owner_t> *buf) {
-  BndInfo out;
-  out.buf = buf->buffer();
-  auto buf_state = buf->GetState();
-  if (buf_state == BufferState::received) {
-    out.buf_allocated = true;
-    PARTHENON_DEBUG_REQUIRE(v->IsAllocated(), "Variable must be allocated to receive");
-  } else if (buf_state == BufferState::received_null) {
-    out.buf_allocated = false;
-  } else {
-    PARTHENON_FAIL("Buffer should be in a received state.");
-  }
+ProResInfo ProResInfo::GetInteriorProlongate(std::shared_ptr<MeshBlock> pmb,
+                                             std::shared_ptr<Variable<Real>> v) {
+  ProResInfo out;
+
   out.allocated = v->IsAllocated();
+  if (!out.allocated) return out;
 
   int Nv = v->GetDim(4);
   int Nu = v->GetDim(5);
   int Nt = v->GetDim(6);
 
-  using TE = TopologicalElement;
-  std::vector<TopologicalElement> elements = {TE::CC};
-  if (v->IsSet(Metadata::Face)) elements = {TE::F1, TE::F2, TE::F3};
-  if (v->IsSet(Metadata::Edge)) elements = {TE::E1, TE::E2, TE::E3};
-  if (v->IsSet(Metadata::Node)) elements = {TE::NN};
+  int mylevel = pmb->loc.level();
+  out.coords = pmb->coords;
+
+  if (pmb->pmr) out.coarse_coords = pmb->pmr->GetCoarseCoords();
+
+  out.fine = v->data.Get();
+  out.coarse = v->coarse_s.Get();
+  NeighborBlock nb;
+  // Make the neighbor block coincide with this block
+  nb.SetNeighbor(pmb->loc, Globals::my_rank, mylevel, 0, 0, 0, 0, 0,
+                 NeighborConnect::none, 0, 0);
+  nb.ownership = block_ownership_t(true);
+
+  auto elements = v->GetTopologicalElements();
+  out.ntopological_elements = elements.size();
+  for (auto el : {TE::CC, TE::F1, TE::F2, TE::F3, TE::E1, TE::E2, TE::E3, TE::NN})
+    out.idxer[static_cast<int>(el)] =
+        CalcIndices(nb, pmb, el, IndexRangeType::Exterior, true, {Nt, Nu, Nv});
+  out.refinement_op = RefinementOp_t::Prolongation;
+  return out;
+}
+
+ProResInfo ProResInfo::GetSend(std::shared_ptr<MeshBlock> pmb, const NeighborBlock &nb,
```

Comment:
`GetSend`, `GetSet`, etc have a lot of the same boiler plate... I wonder if it could be unified? And if so, I wonder if it's worth it? It might just make things less clear...

Reply:
I thought about it a little, but I didn't see an obvious way to unify them. I also think I like that the interface is clear about what type of object you are getting.

Reply:
Yeah that's fine with me. I *think* things could be unified but it feels like it comes at the cost of clarity.

Reply:
Agreed that there is probably some unification that could occur, but I think I am going to leave that be for now (since other PRs based on this one would probably require some complex merging to change things in this PR at this point).